docs/architecture/adr-001-umee-module.md

```diff
@@ -0,0 +1,90 @@
+via IBC. ATOM tokens deposited into the Umee pool will mint a derivative meToken
+in a one-to-one ratio, i.e. one ATOM mints one meToken derivative.
+
+Validators on the Umee network will take these deposited ATOM tokens and delegate
+them to a set of governance-controlled validators on the ATOM source chain,
+e.g. the Cosmos Hub.
+
+A user that sent ATOM tokens to the Umee network can then take their derivative
+meTokens and send them to Ethereum via a bridge where a synthetic ERC-20 version
+of the meTokens are minted. Once sent, the meTokens are locked in Umee and the
+user can then freely trade and operate within Ethereum's DeFi ecosystem with the
+synthetic ERC-20 meTokens.
+
+## Decision
+
+To support users sending ATOM tokens to the Umee network via IBC, we will create
+a `x/umee` module that contains a dedicated `ModuleAccount` that represents the
+ATOM pool. In addition, we will implement a [custom IBC module](https://github.com/cosmos/ibc-go/blob/v1.0.0-alpha1/docs/custom.md).
+The custom IBC module will handle receiving custom packets on a custom Umee port.
+The custom module will be nearly identical to the ICS-20 module, except that it
+will perform the following:
+
+- Automatically send source ATOM tokens to the dedicated `ModuleAccount` pool
+  address instead of a user supplied address.
+- Mark the transfer's amount in a pending queue, `PendingDelegationQueue`, to be
+  later used for delegation on the source chain.
+- Track the sender and amount in the module's state.
+- Mint derivative meTokens and send them to the source sender's Umee account.
+
+At the beginning of a staking epoch, E<sub>1</sub> in `BeginBlock`, the `x/umee`
+module will perform the following:
+
+- Move the funds from the `PendingDelegationQueue` into another queue,
+  `PrevPendingDelegationQueue`. This queue is locked as funds cannot be sent to
+  it. In addition, these are the funds that will be delegated to the source chain
+  during the current staking epoch.
+- Clear out `PendingDelegationQueue`.
+
+The validators in the current staking epoch, E<sub>1</sub>, will then construct,
+sign, and broadcast multisig transactions to the Umee network. Once enough
+transactions are received by the `x/umee` module, specifically 2/3 of total
+staking weight, the `x/umee` module will store the constructed multisig transaction
+and emit an event signalling a delegation transaction can be made to the source
+chain.
+
+A separate relayer process will not only be responsible for relaying IBC packets
+to and from the source chain and Umee, but it will also be responsible for
+listening for these delegation events, and broadcasting the constructed multisig
+transaction to the source chain.
+
+Any new transfers during E<sub>1</sub> will be processed in the same manner during
+E<sub>2</sub>.
+
+## Consequences
+
+### Positive
+
+- Allows for cross-chain delegations in a decentralized manner.
+
+### Negative
```

Comment:
I assume that the validator set can't be that large because of the multisig?

Reply:
what do you think the limitations in size would be? could it feasibly expand in the future?

Reply:
Cosmos does not make use of cryptographic threshold signatures. This means for every m-of-n multisig there is m signatures. This would likely result in some big performance implications for consensus but ultimately it depends how often we are updating the delegation allocations of the multisig as well

Here is a [reference implementation](https://github.com/ZenGo-X/multi-party-eddsa/wiki/Aggregated-Ed25519-Signatures) for cryptographic threshold ed25519

Reply:
Yes, the SDK would need to support this in order for us to use threshold signatures.